### ISSUE TITLE
registrar.warp_and_save_slides function not producing output

### ISSUE THREAD

**Issue by melanieloth** [MathOnco/valis#86]
Hello,

I am trying to save the 24 images that I aligned using Valis. Valis successfully runs the registration step and gives the expected output. 

```
registrar = registration.Valis(slide_src_dir, results_dst_dir, imgs_ordered=True)
rigid_registrar, non_rigid_registrar, error_df = registrar.register()
```

The overlaps look great, so I would like to save them. But I get the following output when running `registrar.warp_and_save_slides(registered_slide_dst_dir, crop="overlap")`

```
/home/mloth/miniconda3/envs/valis_wsi/lib/python3.9/site-packages/numba/core/decorators.py:262: NumbaDeprecationWarning: ^[[1mnumb$
 warnings.warn(msg, NumbaDeprecationWarning)
/home/mloth/miniconda3/envs/valis_wsi/lib/python3.9/site-packages/valis/affine_optimizer.py:311: NumbaDeprecationWarning: The 'nop$
 @nba.jit()
/home/mloth/miniconda3/envs/valis_wsi/lib/python3.9/site-packages/valis/affine_optimizer.py:320: NumbaDeprecationWarning: The 'nop$
 @nba.jit()
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
^M0it [00:00, ?it/s]^M0it [00:00, ?it/s]
```

No output is produced.

**Comment by cdgatenbee** [comment]
Hi @melanieloth,

Sorry you've run into this issue, it's a strange one. I have noticed that sometimes things stop getting printed to the terminal after registration, but haven't yet pinpointed the cause. Regarding the slides not being saved, could you please restart your Python session, load the registrar (saved in the `.pickle` file located in `dst_dir/data/`) and try saving the individual slides in a loop, like this:

```
import pathlib
from valis import registration, slide_io

registrar = registration.load_registrar(path_to_registrar)
pathlib.Path(registered_slide_dst_dir).mkdir(exist_ok=True, parents=True)
for slide_name, slide_obj in registrar.slide_dict.items():
    dst_f = os.path.join(registered_slide_dst_dir, f"{slide_name}.ome.tiff")
    slide_obj.warp_and_save_slide(dst_f=dst_f, crop="overlap")
```

If it is working, you should see output like this:

```
saving registered_slides/401427_BrdU.ome.tiff (22541 x 16509 and 3 channels)

[====================================================================================================] 100.0% in 16.62 seconds
Complete

saving /registered_slides/399261_H&E.ome.tiff (22541 x 16509 and 3 channels)

[====================================================================================================] 100.0% in 15.832 seconds
Complete

saving registered_slides/404718_Trichrome.ome.tiff (22541 x 16509 and 3 channels)

[====================================================================================================] 100.0% in 16.052 seconds
Complete
```

If it isn't working, hopefully there will be some error messages this time that will clarify what is going wrong. Either way, please try it out and let me know how it goes.

Best,

-Chandler

**Comment by melanieloth** [comment]
Thank you for the suggestion. After running the loop, I get the following errors:

```
(process:7227): GLib-GObject-CRITICAL **: 11:10:11.464: value "49936" of type 'gint' is invalid or out of range for property 'tile-width' of type 'gint'

(process:7227): GLib-GObject-CRITICAL **: 11:10:11.464: value "49936" of type 'gint' is invalid or out of range for property 'tile-height' of type 'gint'
Traceback (most recent call last):
  File "/home/mloth/MultiplexRegistration/valisTestScript.py", line 13, in <module>
    slide_obj.warp_and_save_slide(dst_f=dst_f, crop="overlap")
  File "/home/mloth/miniconda3/envs/valis_wsi/lib/python3.9/site-packages/valis/valtils.py", line 121, in wrapper
    return f(*args, **kwargs)
  File "/home/mloth/miniconda3/envs/valis_wsi/lib/python3.9/site-packages/valis/registration.py", line 952, in warp_and_save_slide
    ome_xml_obj = slide_io.update_xml_for_new_img(current_ome_xml_str=slide_meta.original_xml,
  File "/home/mloth/miniconda3/envs/valis_wsi/lib/python3.9/site-packages/valis/valtils.py", line 121, in wrapper
    return f(*args, **kwargs)
  File "/home/mloth/miniconda3/envs/valis_wsi/lib/python3.9/site-packages/valis/slide_io.py", line 2279, in update_xml_for_new_img
    temp_new_ome = create_ome_xml(new_xyzct, bf_dtype, is_rgb, pixel_physical_size_xyu, channel_names, colormap=colormap)
  File "/home/mloth/miniconda3/envs/valis_wsi/lib/python3.9/site-packages/valis/valtils.py", line 121, in wrapper
    return f(*args, **kwargs)
  File "/home/mloth/miniconda3/envs/valis_wsi/lib/python3.9/site-packages/valis/slide_io.py", line 2203, in create_ome_xml
    channels = [create_channel(i, name=channel_names[i], color=colormap[channel_names[i]]) for i in range(c)]
  File "/home/mloth/miniconda3/envs/valis_wsi/lib/python3.9/site-packages/valis/slide_io.py", line 2203, in <listcomp>
    channels = [create_channel(i, name=channel_names[i], color=colormap[channel_names[i]]) for i in range(c)]
KeyError: 'None'
```

**Comment by cdgatenbee** [comment]
Ooops! Based on the error message, I had assumed all of your images were multiplexed non-RGB images, and didn't include what to do when they are RGB. I also updated the functions to rename channel names that are `'None'` or `None` (previously only checked for the latter). I tested the following code with both IF and RGB images so hopefully it works for you as well:

```
import ome_types


def get_colormap(slide_obj):

    if slide_obj.is_rgb:
        colormap = {None: (255, 255, 255)}

    else:
        nc = slide_obj.reader.metadata.n_channels
        channel_colors = slide_tools.get_matplotlib_channel_colors(nc)
        default_colormap = {slide_obj.reader.metadata.channel_names[i]: channel_colors[i] for i in range(nc)}

        try:
            # Try to get original colors
            og_ome = ome_types.from_xml(slide_obj.reader.metadata.original_xml, parser="xmlschema")
            ome_img = og_ome.images[slide_obj.reader.metadata.series]
            colormap = {c.name: c.color.as_rgb_tuple() for c in ome_img.pixels.channels}
            all_rgb = set(list(colormap.values()))
            nc = len(ome_img.pixels.channels)
            if len(all_rgb) < nc:
                # Channels do not have unique colors
                colormap = default_colormap
        except Exception as e:
            print(e)
            # Can't get original colors
            colormap = default_colormap

    return colormap


def check_channel_names(slide_obj):
    if not slide_obj.is_rgb:
        nc = slide_obj.reader.metadata.n_channels
        default_channel_names = [f"C{i+1}" for i in range(nc)]

        if slide_obj.reader.metadata.channel_names is None:
            channel_names = default_channel_names
        else:
            channel_names = [slide_obj.reader.metadata.channel_names[i] if
                            (slide_obj.reader.metadata.channel_names[i] is not None and slide_obj.reader.metadata.channel_names[i] != "None")
                             else default_channel_names[i] for i in range(nc)]

        slide_obj.reader.metadata.channel_names = channel_names


for slide_name, slide_obj in registrar.slide_dict.items():

    # Make sure no channel names are "None" or None
    check_channel_names(slide_obj)
    # Create colormap using channel names
    cmap = get_colormap(slide_obj)

    dst_f = os.path.join(registered_slide_dst_dir, f"{slide_name}.ome.tiff")
    slide_obj.warp_and_save_slide(dst_f=dst_f, crop="overlap", colormap=cmap)


```

Please try that out and let me know how it goes.

Best,
-Chandler

**Comment by melanieloth** [comment]
Hi Chandler,

Here is my most recent error: 

```
(process:72880): GLib-GObject-CRITICAL **: 14:20:49.541: value "49936" of type 'gint' is invalid or out of range for property 'tile-width' of type 'gint'

(process:72880): GLib-GObject-CRITICAL **: 14:20:49.541: value "49936" of type 'gint' is invalid or out of range for property 'tile-height' of type 'gint'
Traceback (most recent call last):
File "/home/mloth/MultiplexRegistration/valisTestScript.py", line 65, in <module>
    slide_obj.warp_and_save_slide(dst_f=dst_f, crop="overlap", colormap=cmap)
  File "/home/mloth/miniconda3/envs/valis_wsi/lib/python3.9/site-packages/valis/valtils.py", line 121, in wrapper
    return f(*args, **kwargs)
  File "/home/mloth/miniconda3/envs/valis_wsi/lib/python3.9/site-packages/valis/registration.py", line 952, in warp_and_save_slide
    ome_xml_obj = slide_io.update_xml_for_new_img(current_ome_xml_str=slide_meta.original_xml,
  File "/home/mloth/miniconda3/envs/valis_wsi/lib/python3.9/site-packages/valis/valtils.py", line 121, in wrapper
    return f(*args, **kwargs)
  File "/home/mloth/miniconda3/envs/valis_wsi/lib/python3.9/site-packages/valis/slide_io.py", line 2279, in update_xml_for_new_img
    temp_new_ome = create_ome_xml(new_xyzct, bf_dtype, is_rgb, pixel_physical_size_xyu, channel_names, colormap=colormap)
  File "/home/mloth/miniconda3/envs/valis_wsi/lib/python3.9/site-packages/valis/valtils.py", line 121, in wrapper
    return f(*args, **kwargs)
  File "/home/mloth/miniconda3/envs/valis_wsi/lib/python3.9/site-packages/valis/slide_io.py", line 2203, in create_ome_xml
    channels = [create_channel(i, name=channel_names[i], color=colormap[channel_names[i]]) for i in range(c)]
  File "/home/mloth/miniconda3/envs/valis_wsi/lib/python3.9/site-packages/valis/slide_io.py", line 2203, in <listcomp>
    channels = [create_channel(i, name=channel_names[i], color=colormap[channel_names[i]]) for i in range(c)]
KeyError: 'C1'


```

**Comment by cdgatenbee** [comment]
Hi @melanieloth, 
It seems this error is probably related to the one described in [Issue 78](https://github.com/MathOnco/valis/issues/78), and I think has something to do with a mismatch between channel names in the metadata and the keys in the colormap dictionary. Not being able to recreate the issue, I'm still not certain how this is happening or what is causing the problem, but it seems that a simple fix is to just save the images without any colormap. Software like QuPath will add the colors if none are provided, so I don't think this will cause downstream issues. Would you mind testing if the code below fixes this issue for you? If there is a mismatch between channel names and the colormap, or you set `colormap=None` in `warp_and_save_slide`, then the image is saved without channel colors in the metadata. It should also print a message about which channels have the mismatch (if any). 

```
import pathlib
import os
import ome_types
import xml.etree.ElementTree as elementTree
import traceback
import numpy as np
from valis import registration, slide_io, slide_tools


CMAP_AUTO = "auto"


def get_colormap(channel_names, is_rgb, series=0, original_xml=None):

    if is_rgb:
        colormap = {None: (255, 255, 255)}

    else:
        nc = len(channel_names)
        channel_colors = slide_tools.get_matplotlib_channel_colors(nc)
        default_colormap = {channel_names[i]: channel_colors[i] for i in range(nc)}

        if original_xml is not None:
            try:
                # Try to get original colors
                og_ome = ome_types.from_xml(original_xml, parser="xmlschema")
                ome_img = og_ome.images[series]
                colormap = {c.name: c.color.as_rgb_tuple() for c in ome_img.pixels.channels}
                all_rgb = set(list(colormap.values()))
                nc = len(ome_img.pixels.channels)
                if len(all_rgb) < nc:
                    # Channels do not have unique colors
                    colormap = default_colormap
            except Exception as e:
                print(e)
                # Can't get original colors
                colormap = default_colormap
        else:
            colormap = default_colormap

    return colormap


def check_channel_names(channel_names, is_rgb):

    if is_rgb:
        return None

    nc = len(channel_names)
    default_channel_names = [f"C{i+1}" for i in range(nc)]

    if channel_names is None:
        updated_channel_names = default_channel_names
    else:
        updated_channel_names = [channel_names[i] if
                                 (channel_names[i] is not None and channel_names[i] != "None")
                                 else default_channel_names[i] for i in range(nc)]

    renamed_channels = set(updated_channel_names) - set(channel_names)
    if len(renamed_channels) > 0:
        msg = f"some non-RGB channel names were `None`. Renamed channels are: {sorted(list(renamed_channels))}"
        print(msg)

    return updated_channel_names


def create_ome_xml(shape_xyzct, bf_dtype, is_rgb, pixel_physical_size_xyu=None, channel_names=None, colormap=CMAP_AUTO):
    """Create new ome-xmml object

    Parameters
    -------
    shape_xyzct : tuple of int
        XYZCT shape of image

    bf_dtype : str
        String format of Bioformats datatype

    is_rgb : bool
        Whether or not the image is RGB

    pixel_physical_size_xyu : tuple, optional
        Physical size per pixel and the unit.

    channel_names : list, optional
        List of channel names.

    colormap : dict, optional
        Dictionary of channel colors, where the key is the channel name, and the value the color as rgb255.
        If None (default), the channel colors from `current_ome_xml_str` will be used, if available.

    Returns
    -------
    new_ome : ome_types.model.OME
        ome_types.model.OME object containing ome-xml metadata

    """

    x, y, z, c, t = shape_xyzct
    new_ome = ome_types.OME()
    new_img = ome_types.model.Image(
        id="Image:0",
        pixels=ome_types.model.Pixels(
            id="Pixels:0",
            size_x=x,
            size_y=y,
            size_z=z,
            size_c=c,
            size_t=t,
            type=bf_dtype,
            dimension_order='XYZCT',
            metadata_only=True
        )
    )

    if pixel_physical_size_xyu is not None:
        phys_x, phys_y, phys_u = pixel_physical_size_xyu
        new_img.pixels.physical_size_x = phys_x
        new_img.pixels.physical_size_x_unit = phys_u
        new_img.pixels.physical_size_y = phys_y
        new_img.pixels.physical_size_y_unit = phys_u

    if is_rgb:
        rgb_channel = ome_types.model.Channel(id='Channel:0:0', samples_per_pixel=3)
        new_img.pixels.channels = [rgb_channel]

    else:
        updated_channel_names = check_channel_names(channel_names, is_rgb)
        if colormap == CMAP_AUTO:
            colormap = get_colormap(updated_channel_names, is_rgb)

        if colormap is not None:
            try:
                channels = [slide_io.create_channel(i, name=updated_channel_names[i], color=colormap[updated_channel_names[i]]) for i in range(c)]
            except KeyError as e:
                msg = f"Mismatch between channel names and keys in colormap. Cannot find channel name {e} in colormap"
                if colormap is not None:
                    msg += f", which has keys: {list(colormap.keys())}"
                msg += ". Saving without colormap. To avoid this error, please provide valid colormap, or set colormap=None."
                print(msg)
                channels = [slide_io.create_channel(i, name=updated_channel_names[i]) for i in range(c)]

        else:
            channels = [slide_io.create_channel(i, name=updated_channel_names[i]) for i in range(c)]

        new_img.pixels.channels = channels

    new_ome = ome_types.model.OME()
    new_ome.images.append(new_img)

    return new_ome


def update_xml_for_new_img(current_ome_xml_str, new_xyzct, bf_dtype, is_rgb, series, pixel_physical_size_xyu=None, channel_names=None, colormap=CMAP_AUTO):
    """Update dimensions ome-xml metadata

    Used to create a new ome-xmlthat reflects changes in an image, such as its shape

    If `current_ome_xml_str` is invalid or None, a new ome-xml will be created

    Parameters
    -------
    current_ome_xml_str : str
        ome-xml string that needs to be updated

    new_xyzct : tuple of int
        XYZCT shape of image

    bf_dtype : str
        String format of Bioformats datatype

    is_rgb : bool
        Whether or not the image is RGB

    pixel_physical_size_xyu : tuple, optional
        Physical size per pixel and the unit.

    channel_names : list, optional
        List of channel names.

    colormap : dict, optional
        Dictionary of channel colors, where the key is the channel name, and the value the color as rgb255.
        If None (default), the channel colors from `current_ome_xml_str` will be used, if available.
        If None, and there are no channel colors in the `current_ome_xml_str`, then no colors will be added

    Returns
    -------
    new_ome : ome_types.model.OME
        ome_types.model.OME object containing ome-xml metadata

    """

    og_valid_xml = True
    og_ome = None
    if current_ome_xml_str is not None:
        try:
            elementTree.fromstring(current_ome_xml_str)
            og_ome = ome_types.from_xml(current_ome_xml_str, parser="xmlschema")
        except elementTree.ParseError as e:
            traceback_msg = traceback.format_exc()
            # msg = "xml in original file is invalid or missing. Will create one"
            print(traceback_msg)
            # valtils.print_warning(msg, traceback_msg=traceback_msg)
            og_valid_xml = False

    else:
        og_valid_xml = False

    updated_channel_names = check_channel_names(channel_names, is_rgb)
    if colormap == CMAP_AUTO:
        colormap = get_colormap(updated_channel_names, is_rgb=is_rgb, series=series, original_xml=current_ome_xml_str)

    temp_new_ome = create_ome_xml(shape_xyzct=new_xyzct, bf_dtype=bf_dtype, is_rgb=is_rgb,
                                  pixel_physical_size_xyu=pixel_physical_size_xyu,
                                  channel_names=updated_channel_names, colormap=colormap)

    if og_valid_xml:
        new_ome = og_ome.copy()
        new_ome.images = temp_new_ome.images
    else:
        new_ome = temp_new_ome

    return new_ome


def warp_and_save_slide(slide_obj, dst_f, level=0, non_rigid=True,
                        crop=True, src_f=None,
                        channel_names=None,
                        colormap=CMAP_AUTO,
                        interp_method="bicubic",
                        tile_wh=None, compression="lzw",
                        Q=100,
                        reader=None):

    """Warp and save a slide

    Slides will be saved in the ome.tiff format.

    Parameters
    ----------
    dst_f : str
        Path to were the warped slide will be saved.

    level : int
        Pyramid level to be warped

    non_rigid : bool, optional
        Whether or not to conduct non-rigid warping. If False,
        then only a rigid transformation will be applied. Default is True

    crop: bool, str
        How to crop the registered images. If `True`, then the same crop used
        when initializing the `Valis` object will be used. If `False`, the
        image will not be cropped. If "overlap", the warped slide will be
        cropped to include only areas where all images overlapped.
        "reference" crops to the area that overlaps with the reference image,
        defined by `reference_img_f` when initializing the `Valis object`.

    channel_names : list, optional
        List of channel names. If None, then Slide.reader
        will attempt to find the channel names associated with `src_f`.

    colormap : dict, optional
        Dictionary of channel colors, where the key is the channel name, and the value the color as rgb255.
        If None (default), the channel colors from `current_ome_xml_str` will be used, if available.
        If None, and there are no channel colors in the `current_ome_xml_str`, then no colors will be added

    src_f : str, optional
        Path of slide to be warped. If None (the default), Slide.src_f
        will be used. Otherwise, the file to which `src_f` points to should
        be an alternative copy of the slide, such as one that has undergone
        processing (e.g. stain segmentation), has a mask applied, etc...

    interp_method : str
        Interpolation method used when warping slide. Default is "bicubic"

    tile_wh : int, optional
        Tile width and height used to save image

    compression : str
        Compression method used to save ome.tiff . Default is lzw, but can also
        be jpeg or jp2k. See pyips for more details.

    Q : int
        Q factor for lossy compression

    pyramid : bool
        Whether or not to save an image pyramid.
    """
    if src_f is None:
        src_f = slide_obj.src_f

    warped_slide = slide_obj.warp_slide(level=level, non_rigid=non_rigid,
                                    crop=crop,
                                    interp_method=interp_method,
                                    src_f=src_f)

    # Get ome-xml #
    if reader is None:
        if src_f != slide_obj.src_f:
            slide_reader_cls = slide_io.get_slide_reader(src_f)
            reader = slide_reader_cls(src_f)
        else:
            reader = slide_obj.reader

    slide_meta = reader.metadata
    if slide_meta.pixel_physical_size_xyu[2] == slide_io.PIXEL_UNIT:
        px_phys_size = None
    else:
        px_phys_size = reader.scale_physical_size(level)

    if channel_names is None:
        channel_names = slide_meta.channel_names

    bf_dtype = slide_io.vips2bf_dtype(warped_slide.format)
    out_xyczt = slide_io.get_shape_xyzct((warped_slide.width, warped_slide.height), warped_slide.bands)
    ome_xml_obj = update_xml_for_new_img(current_ome_xml_str=slide_meta.original_xml,
                                                    new_xyzct=out_xyczt,
                                                    bf_dtype=bf_dtype,
                                                    is_rgb=slide_obj.is_rgb,
                                                    series=slide_obj.series,
                                                    pixel_physical_size_xyu=px_phys_size,
                                                    channel_names=channel_names,
                                                    colormap=colormap
                                                    )

    ome_xml = ome_xml_obj.to_xml()
    if tile_wh is None:
        tile_wh = slide_meta.optimal_tile_wh
        if level != 0:
            down_sampling = np.mean(slide_meta.slide_dimensions[level]/slide_meta.slide_dimensions[0])
            tile_wh = int(np.round(tile_wh*down_sampling))
            tile_wh = tile_wh - (tile_wh % 16)  # Tile shape must be multiple of 16
            if tile_wh < 16:
                tile_wh = 16
            if np.any(np.array(out_xyczt[0:2]) < tile_wh):
                tile_wh = min(out_xyczt[0:2])

    slide_io.save_ome_tiff(warped_slide, dst_f=dst_f, ome_xml=ome_xml,
                           tile_wh=tile_wh, compression=compression,
                           Q=Q)



for slide_name, slide_obj in registrar.slide_dict.items():

    dst_f = os.path.join(registered_slide_dst_dir, f"{slide_name}.ome.tiff")
    warp_and_save_slide(slide_obj, dst_f=dst_f, crop="overlap", level=0)

```

Please try that and let me know if it works for you. Thank you in advance for your patience.

Best,
-Chandler

**Comment by smith6jt-cop** [comment]
I had this problem and this last script worked for me!  Thank you so much!

**Comment by cdgatenbee** [comment]
Great to hear this fixed the issue for you, @smith6jt-cop! I'll have this fix in the next update (1.0.3), which I hope to have out in the next few days.

Best,
-Chandler
